File: Binary_Search/Binary_Search.py
# Binary Search in Python

def binary_search(arr, target):
    left = 0  # Index of the leftmost element of the array.
    right = len(arr) - 1  # Index of the rightmost element of the array.

    while left <= right:
        mid = (left + right) // 2  # Calculate the midpoint of the array.

        if arr[mid] == target:
            return mid  # Return the index of the target element.
        elif arr[mid] < target:
            left = mid + 1  # Update the left pointer.
        else:
            right = mid - 1  # Update the right pointer.

    return -1  # Return -1 if the target element is not found.


#Example
arr = [1, 3, 5, 7, 9, 11, 13, 15]
target = 7
result = binary_search(arr, target)
if result != -1:
    print(f"Element {target} is present at index {result}.")
else:
    print(f"Element {target} is not present in the array.")


# This code is contributed by Erfan Shafiee

File: Binary_Search/Binary_Search.go
// Binary Search in go
package Binary_Search

import (
	"fmt"
)

// binarySearch performs binary search on a sorted array.
// It returns the index of the target value if found, otherwise returns -1.
func binarySearch(arr []int, target int) int {
	// Initialize the start and end pointers
	start, end := 0, len(arr)-1

	// Loop until the pointers cross
	for start <= end {
		// Calculate the middle index
		mid := start + (end-start)/2

		// Check if the target is present at mid
		if arr[mid] == target {
			return mid // Target found
		}

		// If target is greater, ignore the left half
		if arr[mid] < target {
			start = mid + 1
		} else {
			// If target is smaller, ignore the right half
			end = mid - 1
		}
	}

	// Target not found
	return -1
}

func main() {
	// Example sorted array
	arr := []int{2, 3, 4, 10, 40}
	target := 10

	// Call binarySearch and store the result
	result := binarySearch(arr, target)

	// Print the result
	if result != -1 {
		fmt.Printf("Element found at index %d\n", result)
	} else {
		fmt.Println("Element not found in the array")
	}
}

// This code is contributed by Erfan Shafiee